app/lib/service/entrypoint/_messages.dart

```diff
@@ -0,0 +1,172 @@
+        );
+      case 'reply':
+        return ReplyMessage._(
+          error: inner['error'] as String?,
+          result: inner['result'],
+        );
+      case 'debug':
+        return DebugMessage(inner['text'] as String);
+    }
+    throw FormatException('Unknown key: "$key".');
+  }
+
+  Map<String, dynamic> encodeAsJson();
+}
+
+/// Initializing message send from the controller isolate to the new one.
+class EntryMessage extends Message {
+  final SendPort protocolSendPort;
+  final SendPort aliveSendPort;
+
+  EntryMessage({
+    required this.protocolSendPort,
+    required this.aliveSendPort,
+  });
+
+  @override
+  Map<String, dynamic> encodeAsJson() {
+    return {
+      'entry': {
+        'protocolSendPort': protocolSendPort,
+        'aliveSendPort': aliveSendPort,
+      },
+    };
+  }
+}
+
+/// Message sent from the isolate to indicate that it is ready with the initialization.
+class ReadyMessage extends Message {
+  final SendPort? requestSendPort;
+
+  ReadyMessage({
+    this.requestSendPort,
+  });
+
+  @override
+  Map<String, dynamic> encodeAsJson() {
+    return {
+      'ready': {
+        'requestSendPort': requestSendPort,
+      },
+    };
+  }
+}
+
+/// Message sent to a different isolate group for processing.
+class RequestMessage extends Message {
+  /// The `kind` of the target isolate group.
+  final String kind;
+
+  /// The payload (request) object.
+  final Object payload;
+
+  /// The port where we expect the respnse.
+  final SendPort replyPort;
+
+  RequestMessage(this.kind, this.payload, this.replyPort);
+
+  @override
+  Map<String, dynamic> encodeAsJson() {
+    return {
+      'request': {
+        'kind': kind,
+        'payload': payload,
+        'replyPort': replyPort,
+      },
+    };
+  }
+}
+
+/// Message sent as a reply to a [RequestMessage].
+class ReplyMessage extends Message {
+  final String? error;
+  final Object? result;
+
+  ReplyMessage._({
+    required this.error,
+    required this.result,
+  });
+  ReplyMessage.error(String message)
+      : error = message,
+        result = null;
+  ReplyMessage.result(Object value)
+      : result = value,
+        error = null;
+
+  bool get isError => error != null;
+
+  @override
+  Map<String, dynamic> encodeAsJson() {
+    return {
+      'reply': {
+        if (error != null) 'error': error,
+        if (result != null) 'result': result,
+      }
+    };
+  }
+}
+
+/// Message sent from the isolate with arbitrary text.
+class DebugMessage extends Message {
+  final String text;
+
+  DebugMessage(this.text);
+
+  @override
+  Map<String, dynamic> encodeAsJson() {
+    return {
+      'debug': {
+        'text': text,
+      },
+    };
+  }
+}
+
+/// Wraps the request/reply message sending logic.
+Future<Object?> handleRequestReply({
+  required SendPort requestSendPort,
+  required String kind,
+  required Object payload,
+  required Duration timeout,
+}) async {
+  final replyRecievePort = ReceivePort();
```

Comment:
I think we need to close this in a `finally`, otherwise we're leaking ports.

I'm not sure how well GC'ing works on ports, it's probably one of things that it's helpful to close.

Especially, since you're not cancelling the listening operation. It's unlikely that the GC will ever be able to see that all instances of the [SendPort] have been GC'ed. I'm just guessing here. But these can be sent to FFI stuff, so our isolate will probably be unable to die naturally, it'll always sit around waiting for a message on the port, if a timeout happens.

Reply:
Good point, let's do it explicitly, although, implicitly we are already doing it I think:

"A receive port is closed by canceling its subscription." + "Internally the method cancels its subscription after the first element." (`RecievePort.first`)

Reply:
If the message never comes and there is a timeout, then `RecievePort.first` will never be resolved, and it'll never cause the subscription to be canceled.